### DISCUSSION TITLE
React Server Components and Realtime

### DISCUSSION THREAD

**Discussion by aretrace** [pocketbase/pocketbase#2057] [Q&A] [answered]
Using Next.js 13 (appDir), how could you use subscriptions in a react server component to (re)render record changes from a collection on the server?

So far in a page.js...
``` javascript
// ..
try {
  await pb.collection('test').subscribe('*', (e) => (
    console.log(e.record.id)
  ))
} catch (err) {
  console.log(err)
}

export default function ServerComponent() {
// ..
```
...gives an initial error:
`
ClientResponseError 0: Something went wrong while processing your request.
    at new ClientResponseError (webpack-internal:///(sc_server)/./node_modules/pocketbase/dist/pocketbase.es.mjs:160:23)
    at RealtimeService.connectErrorHandler (webpack-internal:///(sc_server)/./node_modules/pocketbase/dist/pocketbase.es.mjs:1277:29)
    at eval (webpack-internal:///(sc_server)/./node_modules/pocketbase/dist/pocketbase.es.mjs:1262:36) {
  url: '',
  status: 0,
  response: {},
  isAbort: false,
  originalError: null
}
`
...and proceeds to behave as expected (callback is called for create, update, and delete record operation events). 

The goal here is to simply rerender the component when a sqlite row is changed (create, update, and delete). Pocketbase is one of the surprisingly few methods to achieve this in a node environment but I would like to mix in react server components (if it's even possible / makes any sense). 

Any ideas on how to combine react server components and the realtime capabilities of pocketbase would be immensely helpful (without useState or useEffect as that will make a client side component), as well any ideas/solutions for the general goal as well. I have already ruled out solutions like Realm and TinyBase. I'm looking for a sqlite based solution, another potential tool I found is WatermelonDB (the interaction with RSC is unfamiliar to me as well).

**Comment by ganigeorgiev** [comment]
I'm not familiar with Next.js 13 and I don't know what to recommend/suggest, but I don't think you can rerender a server-side component without calling [`router.refresh()`](https://beta.nextjs.org/docs/data-fetching/mutating) or other similar primitive.

If you haven't already, you can also check https://beta.nextjs.org/docs/rendering/server-and-client-components#when-to-use-server-vs-client-components.

About the initial `ClientResponseError` - it looks strange, I'm not sure what is causing it. Do you have an [`EventSource` polyfill loaded](https://github.com/pocketbase/js-sdk#nodejs-via-npm)? Since the original error may not be json stringifiable, you can try `console.log(err.originalError)` and see if there are any details.

**Reply by aretrace** [reply]
EventSource polyfill is loaded, originalError is `null`. Disabled strict mode and have autoCancellation set to false, the initial error persists.

**Comment by aretrace** [comment, marked as answer]
It turns out that trying to '(re)render' react server components makes no sense. These kinds of components only run once in a server context with their results streamed to the client, they're not capable of maintaining an internal state. However, they are capable of rendering and passing down data to client-side components. Using them for initial data fetching seems to be a primary use case with other benefits stemming from their particular behavior.
